adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		}
+		if content.MetaData.VideoInfo.Width != 0 && content.MetaData.VideoInfo.Height != 0 {
+			adWidth = int64(content.MetaData.VideoInfo.Width)
+			adHeight = int64(content.MetaData.VideoInfo.Height)
+		} else if bidType == openrtb_ext.BidTypeVideo {
+			if opentrb2Imp.Video != nil && opentrb2Imp.Video.W != 0 && opentrb2Imp.Video.H != 0 {
+				adWidth = opentrb2Imp.Video.W
+				adHeight = opentrb2Imp.Video.H
+			}
+		} else {
+			return "", 0, 0, errors.New("extractAdmVideo: cannot get width, height")
+		}
+		duration = getDuration(int64(content.MetaData.VideoInfo.VideoDuration))
+	}
+
+	var adTitle, _ = getDecodeValue(content.MetaData.Title)
+	var adId = content.Contentid
+	var creativeId = content.Contentid
+	var trackingEvents strings.Builder
+	var dspImpTracking2Str = ""
+	var dspClickTracking2Str = ""
+	var errorTracking2Str = ""
+	if content.Monitor != nil {
+		for _, monitor := range content.Monitor {
+			if len(monitor.Url) == 0 {
+				continue
+			}
+			var event = ""
+			switch monitor.EventType {
+			case "vastError":
+				errorTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "vastError")
+			case "imp":
+				dspImpTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "imp")
+			case "click":
+				dspClickTracking2Str = getVastImpClickErrorTrackingUrls(monitor.Url, "click")
+			case "userclose":
+				event = "skip&closeLinear"
+			case "playStart":
+				event = "start"
+			case "playEnd":
+				event = "complete"
+			case "playResume":
+				event = "resume"
+			case "playPause":
+				event = "pause"
+			case "soundClickOff":
+				event = "mute"
+			case "soundClickOn":
+				event = "unmute"
+			default:
+			}
+			if event != "" {
+				if event != "skip&closeLinear" {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, event))
+				} else {
+					trackingEvents.WriteString(getVastEventTrackingUrls(monitor.Url, "skip&closeLinear"))
+				}
+			}
+		}
+	}
+
+	adm = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+		"<VAST version=\"3.0\">" +
+		"<Ad id=\"" + adId + "\"><InLine>" +
+		"<AdSystem>HuaweiAds</AdSystem>" +
+		"<AdTitle>" + adTitle + "</AdTitle>" +
+		errorTracking2Str + dspImpTracking2Str +
+		"<Creatives>" +
+		"<Creative adId=\"" + adId + "\" id=\"" + creativeId + "\">" +
+		"<Linear>" +
+		"<Duration>" + duration + "</Duration>" +
+		"<TrackingEvents>" + trackingEvents.String() + "</TrackingEvents>" +
+		"<VideoClicks>" +
+		"<ClickThrough><![CDATA[" + clickUrl + "]]></ClickThrough>" +
+		dspClickTracking2Str +
+		"</VideoClicks>" +
+		"<MediaFiles>" +
+		"<MediaFile delivery=\"progressive\" type=\"" + mime + "\" width=\"" + strconv.Itoa(int(adWidth)) + "\" " +
+		"height=\"" + strconv.Itoa(int(adHeight)) + "\" scalable=\"true\" maintainAspectRatio=\"true\"> " +
+		"<![CDATA[" + resourceUrl + "]]>" +
+		"</MediaFile>" +
+		"</MediaFiles>" +
+		"</Linear>" +
+		"</Creative>" +
+		"</Creatives>" +
+		"</InLine>" +
+		"</Ad>" +
+		"</VAST>"
```

Comment:
Consider using back tick literal strings here to avoid escaping the double quote.

Reply:
corrected